instance/models/instance.py

```diff
@@ -547,6 +686,78 @@ def studio_url(self):
         """
         return u'{0.protocol}://{0.studio_domain}/'.format(self)
 
+    @property
+    def database_name(self):
+        """
+        The database name used for external databases. Escape all non-ascii characters and truncate to 64 chars, the
+        maximum for mysql
+        """
+        name = self.domain.replace('.', '_')
+        allowed = string.ascii_letters + string.digits + '_'
+        escaped = ''.join(char for char in name if char in allowed)
+        return truncate_name(escaped, length=64)
+
+    @property
+    def mysql_database_name(self):
+        """
+        The mysql database name for this instance
+        """
+        return self.database_name
+
+    @property
+    def mysql_database_names(self):
+        """
+        List of mysql database names
+        """
+        return [self.mysql_database_name]
+
+    @property
+    def mongo_database_name(self):
+        """
+        The name of the main external mongo database
+        """
+        return self.database_name
+
+    @property
+    def forum_database_name(self):
+        """
+        The name of the external database used for forums
+        """
+        return '{0}_forum'.format(self.database_name)
+
+    @property
+    def mongo_database_names(self):
+        """
+        List of mongo database names
+        """
+        return [self.mongo_database_name, self.forum_database_name]
+
+    @staticmethod
+    def on_pre_save(sender, instance, **kwargs):
+        """
+        Set this instance's default field values
+        """
+        super().on_pre_save(sender, instance, **kwargs)
+
+        self = instance
+
+        if self.use_ephemeral_databases is None:
+            self.use_ephemeral_databases = settings.INSTANCE_EPHEMERAL_DATABASES
+        for attr in ('mysql_user', 'mongo_user'):
+            if not getattr(self, attr, None):
+                setattr(self, attr, get_random_string(length=16, allowed_chars=string.ascii_lowercase))
+        for attr in ('mysql_pass', 'mongo_pass'):
+            if not getattr(self, attr, None):
+                setattr(self, attr, get_random_string(length=32))
```

Comment:
We should refrain from filling in the database users and passwords in the case that `use_ephemeral_databases` is `True`.  Otherwise, we will have values for these fields in the database that are completely meaningless and different from the values that are actually used on the deployed instance, which is a bit confusing.